### PR TITLE
SONAR-6912 do not deactivate last organization administrator

### PR DESCRIPTION
**Pull request by simonbrandhof**

### CODE REVIEW

File: server/sonar-server/src/test/java/org/sonar/server/user/ws/ChangePasswordActionTest.java [hunk 1]
```diff
@@ -57,10 +57,11 @@
 
 public class ChangePasswordActionTest {
 
+  private System2 system2 = mock(System2.class);
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/user/ws/ChangePasswordActionTest.java]: no use is actually made of this mock

especially, the property is hidden by a variable with the same name (but storing an instance of `System2`) in method `setUp`

File: server/sonar-server/src/test/java/org/sonar/server/user/ws/CreateActionTest.java [hunk 2]
```diff
@@ -78,7 +79,7 @@
   public void setUp() {
     System2 system2 = new System2();
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/user/ws/CreateActionTest.java]: this variable is hiding the property you defined. I don't think indexer is using the `System2` instance you expect it to be using

File: server/sonar-server/src/test/java/org/sonar/server/user/UserUpdaterTest.java [hunk 3]
```diff
@@ -88,7 +88,7 @@
 
   @Before
   public void setUp() {
-    userIndexer = new UserIndexer(dbClient, es.client());
+    userIndexer = new UserIndexer(system2, dbClient, es.client());
```
Comment (simonbrandhof) [inline server/sonar-server/src/test/java/org/sonar/server/user/UserUpdaterTest.java]: Ticket https://jira.sonarsource.com/browse/SONAR-8390 created.

**Comment by simonbrandhof** [comment]
Bad mocks replaced by System.INSTANCE. It was a bad copy-paste during mass update.

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java [hunk 4]
```diff
@@ -73,33 +72,33 @@ public void handle(Request request, Response response) throws Exception {
...
-    try {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      JsonWriter json = response.newJsonWriter().beginObject();
+      json.name("user");
+      Set<String> groups = Sets.newHashSet();
       UserDto user = dbClient.userDao().selectByLogin(dbSession, login);
-      if (user == null) {
-        throw new NotFoundException(format("User '%s' doesn't exist", login));
-      }
+      checkFound(user, "User '%s' doesn't exist", login);
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java]: why starting to write JSON response and creating a Set if it's for potentially fail when user doesn't exist?

you could move line 96 and 97 to be beginning of the try

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java [hunk 5]
```diff
@@ -101,4 +112,34 @@ private void writeResponse(Response response, String login) {
     }
   }
 
+  private void ensureNotLastAdministrator(DbSession dbSession, UserDto user) {
+    Set<String> organizationUuids = dbClient.authorizationDao().selectOrganizationUuidsOfUserWithGlobalPermission(
+      dbSession, user.getId(), SYSTEM_ADMIN);
+    List<String> problematicOrganizations = new ArrayList<>();
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java]: please move lines 118 to 124 to a function for better readability

File: server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java [hunk 6]
```diff
@@ -101,4 +112,34 @@ private void writeResponse(Response response, String login) {
...
+    for (String organizationUuid : organizationUuids) {
+      int remaining = dbClient.authorizationDao().countUsersWithGlobalPermissionExcludingUser(dbSession, organizationUuid, SYSTEM_ADMIN, user.getId());
+      if (remaining == 0) {
+        problematicOrganizations.add(organizationUuid);
+      }
+    }
+    if (!problematicOrganizations.isEmpty()) {
+      if (problematicOrganizations.size() == 1 && defaultOrganizationProvider.get().getUuid().equals(problematicOrganizations.get(0))) {
+        throw new BadRequestException("User cannot be deactivated as being the last administrator");
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/user/ws/DeactivateAction.java]: not sur of the English of this sentence, maybe @ganncamp can confirm

**Pull request closed without merging**